shared/chat/conversation/giphy/width.js

```diff
@@ -1,10 +1,221 @@
+  return images.reduce((total, i) => {
+    return total + (i.isCompressable() ? 1 : 0)
+  }, 0)
+}
+
+const numExpandables = images => {
+  return images.reduce((total, i) => {
+    return total + (i.isExpandable() ? 1 : 0)
+  }, 0)
+}
+
+const maxCompressPasses = 15
+const maxExpandPasses = 15
+
+// compressRow attempts to compress a candidate row down to a given width
+const compressRow = (totalWidth, row) => {
+  const compressed = []
+  row.forEach(i => {
+    compressed.push(new Image(i.origWidth, i.margin))
+  })
+  for (let pass = 0; pass < maxCompressPasses; pass++) {
+    const totalCompression = groupWidth(compressed) - totalWidth
+    // figure out how much we will compress this pass
+    let imageComp = Math.ceil(totalCompression / numCompressables(compressed))
+    for (let i = 0; i < compressed.length; i++) {
+      compressed[i].compress(imageComp)
+      const newWidth = groupWidth(compressed)
+      if (newWidth <= totalWidth) {
+        // done here, compression has succeeded
+        return compressed
+      }
+      if (newWidth - imageComp < totalWidth) {
+        // if we are going to compress than less than width with the current imageComp, set it to
+        // the remaining about of space.
+        imageComp = newWidth - totalWidth
+      }
+    }
+  }
+  throw new Error('unable to compress')
+}
+
+// expandRow attempt to expand a candidate row into a given width
+const expandRow = (totalWidth, row) => {
+  const expanded = []
+  row.forEach(i => {
+    expanded.push(new Image(i.origWidth, i.margin))
+  })
+  for (let pass = 0; pass < maxExpandPasses; pass++) {
+    const totalExpansion = totalWidth - groupWidth(expanded)
+    const expandables = numExpandables(expanded)
+    if (expandables === 0) {
+      // if we can't expand anything, just return what we have
+      return expanded
+    }
+    let imageExp = Math.floor(totalExpansion / expandables)
+    if (imageExp === 0) {
+      // if this rounds down into 0, then just set to the remaining distance to be expanded into
+      imageExp = totalWidth - groupWidth(expanded)
+    }
+    for (let i = 0; i < expanded.length; i++) {
+      expanded[i].expand(imageExp)
+      if (groupWidth(expanded) >= totalWidth) {
+        return expanded
+      }
+    }
+  }
+  // always return something in the expand case
+  return expanded
+}
+
+const pickRow = (totalWidth, longIndex, longRow, shortRow) => {
+  let compressed
+  const expanded = expandRow(totalWidth, shortRow)
+  try {
+    compressed = compressRow(totalWidth, longRow)
+  } catch (e) {
+    return {
+      index: longIndex - 1,
+      row: expanded,
+    }
+  }
+  // Pick which row is the closest, biasing for compression if both succeed.
+  const compWidth = groupWidth(compressed)
+  const expandWidth = groupWidth(expanded)
+  const compDistance = totalWidth - compWidth
+  let expandDistance = totalWidth - expandWidth
+  if (expandDistance < 0) {
+    expandDistance = 0
+  }
+  if (compDistance < expandDistance) {
```

Comment:
why not `<=` and then we don't need the `else if` or the `else`?

Reply:
👍